### PR TITLE
Add LivingEnchantmentLevelEvent to alter effective enchantment levels

### PR DESCRIPTION
**Pull request by rubensworks**
Adds an event that allows enchantment levels to be modified when they are calculated.
This makes it possible to change these levels based on the player and its environment, rather than always being fixed per ItemStack.

This is useful in cases like:
* Changing fortune level if the player is sneaking.
* Changing fire aspect modifier if the player is in a warm biome.
* Increase looting level if the player is mounted on a certain entity.
* Changing lure level if fishing while it's night.

This PR deprecates (but does not remove) the EnchantmentHelper#getEnchantmentLevel method and adds a player-sensitive one, which is used instead now.

The only part I'm not convinced about is the application of the unbreaking modifier in `ItemStack#attemptDamageItem`: https://github.com/rubensworks/OpenModLoader/blob/440cb9979f61a59a2821690de09108b151a576e9/patches/net/minecraft/item/ItemStack.java.patch#L8
null is passed as entity parameter, because the containing method does not have an entity parameter, and changing that would require some patching in other places which may be undesirable.
It may be better to add an entity-sensitive version of `ItemStack#attemptDamageItem`, but this is probably something for another PR.

A test mod is included that multiplies the fortune level when mining by 10 levels when the player is sneaking.

### CODE REVIEW

**Comment by iLexiconn** [comment]
Can you put this event (along with the other enchant event) in a class called EnchantEvent please? :)

**Comment by rubensworks** [comment]
@iLexiconn Done!
The two enchantments don't have any semantically overlapping fields, so their parent is a pure wrapper class.

File: patches/net/minecraft/enchantment/EnchantmentHelper.java.patch [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+-import net.minecraft.enchantment.EnchantmentData;
++
+ import net.minecraft.entity.Entity;
+ import net.minecraft.entity.EntityLivingBase;
+ import net.minecraft.entity.EnumCreatureAttribute;
+@@ -26,6 +25,7 @@
+ import net.minecraft.util.Util;
+ import net.minecraft.util.WeightedRandom;
+ import net.minecraft.util.math.MathHelper;
++import xyz.openmodloader.event.impl.EnchantmentEvent;
```
Comment (iLexiconn) [inline patches/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Fix this please :p

File: patches/net/minecraft/inventory/ContainerEnchantment.java.patch [hunk 2]
```diff
@@ -1,10 +1,29 @@
...
+-import net.minecraft.inventory.IInventory;
+-import net.minecraft.inventory.InventoryBasic;
+-import net.minecraft.inventory.Slot;
+ import net.minecraft.item.EnumDyeColor;
+ import net.minecraft.item.ItemStack;
+ import net.minecraft.stats.StatList;
+ import net.minecraft.util.SoundCategory;
+ import net.minecraft.util.math.BlockPos;
+ import net.minecraft.world.World;
++import xyz.openmodloader.event.impl.EnchantmentEvent;
```
Comment (iLexiconn) [inline patches/net/minecraft/inventory/ContainerEnchantment.java.patch]: And this

File: patches/net/minecraft/enchantment/EnchantmentHelper.java.patch (hunk 1 above)
Comment (rubensworks) [inline patches/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Great, thanks IDEA....

File: src/main/java/xyz/openmodloader/event/impl/EnchantmentEvent.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    /**
+     * Fired when the player enchants an item. Allows for the cost and enchantments
+     * to be altered, along with the enchanted item itself and the fuel item.
+     */
+    public static class ItemEnchantedEvent extends EnchantmentEvent {
+
+        /**
+         * The player who is enchanting the item.
+         */
+        private final EntityPlayer player;
```
Comment (iLexiconn) [inline src/main/java/xyz/openmodloader/event/impl/EnchantmentEvent.java]: 4 spaces

**Comment by iLexiconn** [comment]
Other than that everything looks good :)

**Comment by iLexiconn** [comment]
Still didn't fix the imports

File: src/main/java/xyz/openmodloader/event/impl/EnchantmentEvent.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+ * Parent class for enchantment related events. All events that fall within this scope
+ * should extend this class. They should also be added as an inner class however
+ * this is not required.
+ */
+public class EnchantmentEvent extends Event {
+    /**
+     * Fired when the player enchants an item. Allows for the cost and enchantments
+     * to be altered, along with the enchanted item itself and the fuel item.
+     */
+    public static class ItemEnchantedEvent extends EnchantmentEvent {
```
Comment (iLexiconn) [inline src/main/java/xyz/openmodloader/event/impl/EnchantmentEvent.java]: Remove 'Event' (same for other event)

**Comment by rubensworks** [comment]
Yeah, I'm still working on it. Having to rebase constantly because new PR's keep getting merged...

**Comment by iLexiconn** [comment]
You shouldn't have to rebase, we're not touching the same code.

**Comment by rubensworks** [comment]
`OMLTestMod` has been changed at least twice since I created this PR ;-)

**Comment by rubensworks** [comment]
@iLexiconn All done now.

**Comment by iLexiconn** [comment]
Thanks!

**Pull request merged**